### ISSUE TITLE
(More) Reproducible Builds with pip-tools

### ISSUE THREAD

**Issue by JimMadge** [the-turing-way/the-turing-way#3099]
### Summary

[pip-tools](https://pip-tools.readthedocs.io/en/latest/) contains commands to help manage pinned version requirements.

It would let us declare a minimal set of requirements in, for example, `requirements.in` and build from that a full set of pinned requirements (including the whole cascade of dependencies) in `requirements.txt`.

That way we can build the book from any commit using the exact dependencies while making updating dependencies easy using `pip-compile`.

### What needs to be done?

- [ ] Move existing `requirements.txt` to `requirements.in`
- [ ] Build full `requirements.txt` with `pip-compile`
- [ ] Add documentation for updating `requirements.txt`

### Who can help?

* 

---

**Comment by JimMadge** [comment]
Agreed, I think `pip-tools` gives you both.

For example,

`requirements.in` (no pinned versions, or any version specifications)

```
pandas
numpy
matplotlib
datascience
folium
jupyter-book
jinja2
```

`pip-compile requirements.in` gives this `requirements.txt`

<details>

```
#
# This file is autogenerated by pip-compile with Python 3.11
# by the following command:
#
#    pip-compile requirements.in
#
alabaster==0.7.13
    # via sphinx
anyio==3.6.2
    # via jupyter-server
appnope==0.1.3
    # via
    #   ipykernel
    #   ipython
argon2-cffi==21.3.0
    # via
    #   jupyter-server
    #   nbclassic
    #   notebook
argon2-cffi-bindings==21.2.0
    # via argon2-cffi
asttokens==2.2.1
    # via stack-data
attrs==21.4.0
    # via
    #   jsonschema
    #   jupyter-cache
    #   markdown-it-py
    #   sphinx-external-toc
babel==2.12.1
    # via sphinx
backcall==0.2.0
    # via ipython
beautifulsoup4==4.12.2
    # via
    #   nbconvert
    #   pydata-sphinx-theme
bleach==6.0.0
    # via nbconvert
branca==0.6.0
    # via
    #   datascience
    #   folium
certifi==2023.5.7
    # via requests
cffi==1.15.1
    # via argon2-cffi-bindings
charset-normalizer==3.1.0
    # via requests
click==8.1.3
    # via
    #   jupyter-book
    #   sphinx-external-toc
colorama==0.4.6
    # via nbdime
comm==0.1.3
    # via ipykernel
cycler==0.11.0
    # via matplotlib
datascience==0.17.6
    # via -r requirements.in
debugpy==1.6.7
    # via ipykernel
decorator==5.1.1
    # via ipython
defusedxml==0.7.1
    # via nbconvert
docutils==0.17.1
    # via
    #   jupyter-book
    #   myst-nb
    #   myst-parser
    #   pybtex-docutils
    #   pydata-sphinx-theme
    #   sphinx
    #   sphinx-togglebutton
    #   sphinxcontrib-bibtex
entrypoints==0.4
    # via nbconvert
executing==1.2.0
    # via stack-data
fastjsonschema==2.17.1
    # via nbformat
folium==0.14.0
    # via
    #   -r requirements.in
    #   datascience
fonttools==4.39.4
    # via matplotlib
gitdb==4.0.10
    # via gitpython
gitpython==3.1.31
    # via nbdime
idna==3.4
    # via
    #   anyio
    #   requests
imagesize==1.4.1
    # via sphinx
importlib-metadata==6.6.0
    # via myst-nb
ipykernel==6.23.1
    # via
    #   ipywidgets
    #   nbclassic
    #   notebook
ipython==8.13.2
    # via
    #   datascience
    #   ipykernel
    #   ipywidgets
    #   jupyter-sphinx
    #   myst-nb
ipython-genutils==0.2.0
    # via
    #   ipywidgets
    #   nbclassic
    #   notebook
ipywidgets==7.7.5
    # via
    #   jupyter-sphinx
    #   myst-nb
jedi==0.18.2
    # via ipython
jinja2==3.1.2
    # via
    #   -r requirements.in
    #   branca
    #   folium
    #   jupyter-book
    #   jupyter-server
    #   myst-parser
    #   nbclassic
    #   nbconvert
    #   nbdime
    #   notebook
    #   sphinx
jsonschema[format-nongpl]==3.2.0
    # via
    #   jupyter-book
    #   jupyter-events
    #   nbformat
jupyter-book==0.13
    # via -r requirements.in
jupyter-cache==0.4.3
    # via myst-nb
jupyter-client==8.2.0
    # via
    #   ipykernel
    #   jupyter-server
    #   nbclassic
    #   nbclient
    #   notebook
jupyter-core==5.3.0
    # via
    #   ipykernel
    #   jupyter-client
    #   jupyter-server
    #   nbclassic
    #   nbconvert
    #   nbformat
    #   notebook
jupyter-events==0.6.3
    # via jupyter-server
jupyter-server==2.5.0
    # via
    #   jupyter-server-mathjax
    #   nbclassic
    #   nbdime
    #   notebook-shim
jupyter-server-mathjax==0.2.6
    # via nbdime
jupyter-server-terminals==0.4.4
    # via jupyter-server
jupyter-sphinx==0.3.2
    # via myst-nb
jupyterlab-pygments==0.2.2
    # via nbconvert
jupyterlab-widgets==1.1.4
    # via ipywidgets
kiwisolver==1.4.4
    # via matplotlib
latexcodec==2.0.1
    # via pybtex
linkify-it-py==1.0.3
    # via jupyter-book
lxml==4.9.2
    # via nbconvert
markdown-it-py==1.1.0
    # via
    #   mdit-py-plugins
    #   myst-parser
markupsafe==2.1.2
    # via
    #   jinja2
    #   nbconvert
matplotlib==3.5.2
    # via
    #   -r requirements.in
    #   datascience
matplotlib-inline==0.1.6
    # via
    #   ipykernel
    #   ipython
mdit-py-plugins==0.2.8
    # via myst-parser
mistune==0.8.4
    # via nbconvert
myst-nb==0.13.2
    # via jupyter-book
myst-parser==0.15.2
    # via myst-nb
nbclassic==1.0.0
    # via notebook
nbclient==0.5.13
    # via
    #   jupyter-cache
    #   nbconvert
nbconvert==6.5.4
    # via
    #   jupyter-server
    #   jupyter-sphinx
    #   myst-nb
    #   nbclassic
    #   notebook
nbdime==3.2.1
    # via jupyter-cache
nbformat==5.8.0
    # via
    #   jupyter-cache
    #   jupyter-server
    #   jupyter-sphinx
    #   myst-nb
    #   nbclassic
    #   nbclient
    #   nbconvert
    #   nbdime
    #   notebook
nest-asyncio==1.5.6
    # via
    #   ipykernel
    #   nbclassic
    #   nbclient
    #   notebook
notebook==6.5.4
    # via widgetsnbextension
notebook-shim==0.2.3
    # via nbclassic
numpy==1.23.2
    # via
    #   -r requirements.in
    #   datascience
    #   folium
    #   matplotlib
    #   pandas
    #   scipy
packaging==23.1
    # via
    #   ipykernel
    #   jupyter-server
    #   matplotlib
    #   nbconvert
    #   plotly
    #   pydata-sphinx-theme
    #   sphinx
pandas==1.4.3
    # via
    #   -r requirements.in
    #   datascience
pandocfilters==1.5.0
    # via nbconvert
parso==0.8.3
    # via jedi
pexpect==4.8.0
    # via ipython
pickleshare==0.7.5
    # via ipython
pillow==9.5.0
    # via matplotlib
platformdirs==3.5.1
    # via jupyter-core
plotly==5.14.1
    # via datascience
prometheus-client==0.16.0
    # via
    #   jupyter-server
    #   nbclassic
    #   notebook
prompt-toolkit==3.0.38
    # via ipython
psutil==5.9.5
    # via ipykernel
ptyprocess==0.7.0
    # via
    #   pexpect
    #   terminado
pure-eval==0.2.2
    # via stack-data
pybtex==0.24.0
    # via
    #   pybtex-docutils
    #   sphinxcontrib-bibtex
pybtex-docutils==1.0.2
    # via sphinxcontrib-bibtex
pycparser==2.21
    # via cffi
pydata-sphinx-theme==0.8.1
    # via sphinx-book-theme
pygments==2.15.1
    # via
    #   ipython
    #   nbconvert
    #   nbdime
    #   sphinx
pyparsing==3.0.9
    # via matplotlib
pyrsistent==0.19.3
    # via jsonschema
python-dateutil==2.8.2
    # via
    #   jupyter-client
    #   matplotlib
    #   pandas
python-json-logger==2.0.7
    # via jupyter-events
pytz==2023.3
    # via pandas
pyyaml==6.0
    # via
    #   jupyter-book
    #   jupyter-events
    #   myst-nb
    #   myst-parser
    #   pybtex
    #   sphinx-book-theme
    #   sphinx-external-toc
pyzmq==25.0.2
    # via
    #   ipykernel
    #   jupyter-client
    #   jupyter-server
    #   nbclassic
    #   notebook
requests==2.31.0
    # via
    #   folium
    #   nbdime
    #   sphinx
rfc3339-validator==0.1.4
    # via jupyter-events
rfc3986-validator==0.1.1
    # via jupyter-events
scipy==1.10.1
    # via datascience
send2trash==1.8.2
    # via
    #   jupyter-server
    #   nbclassic
    #   notebook
six==1.16.0
    # via
    #   asttokens
    #   bleach
    #   jsonschema
    #   latexcodec
    #   pybtex
    #   python-dateutil
    #   rfc3339-validator
smmap==5.0.0
    # via gitdb
sniffio==1.3.0
    # via anyio
snowballstemmer==2.2.0
    # via sphinx
soupsieve==2.4.1
    # via beautifulsoup4
sphinx==4.5.0
    # via
    #   jupyter-book
    #   jupyter-sphinx
    #   myst-nb
    #   myst-parser
    #   pydata-sphinx-theme
    #   sphinx-book-theme
    #   sphinx-comments
    #   sphinx-copybutton
    #   sphinx-design
    #   sphinx-external-toc
    #   sphinx-jupyterbook-latex
    #   sphinx-multitoc-numbering
    #   sphinx-thebe
    #   sphinx-togglebutton
    #   sphinxcontrib-bibtex
sphinx-book-theme==0.3.3
    # via jupyter-book
sphinx-comments==0.0.3
    # via jupyter-book
sphinx-copybutton==0.5.2
    # via jupyter-book
sphinx-design==0.1.0
    # via jupyter-book
sphinx-external-toc==0.2.4
    # via jupyter-book
sphinx-jupyterbook-latex==0.4.7
    # via jupyter-book
sphinx-multitoc-numbering==0.1.3
    # via jupyter-book
sphinx-thebe==0.1.2
    # via jupyter-book
sphinx-togglebutton==0.3.2
    # via
    #   jupyter-book
    #   myst-nb
sphinxcontrib-applehelp==1.0.4
    # via sphinx
sphinxcontrib-bibtex==2.5.0
    # via jupyter-book
sphinxcontrib-devhelp==1.0.2
    # via sphinx
sphinxcontrib-htmlhelp==2.0.1
    # via sphinx
sphinxcontrib-jsmath==1.0.1
    # via sphinx
sphinxcontrib-qthelp==1.0.3
    # via sphinx
sphinxcontrib-serializinghtml==1.1.5
    # via sphinx
sqlalchemy==1.4.48
    # via jupyter-cache
stack-data==0.6.2
    # via ipython
tenacity==8.2.2
    # via plotly
terminado==0.17.1
    # via
    #   jupyter-server
    #   jupyter-server-terminals
    #   nbclassic
    #   notebook
tinycss2==1.2.1
    # via nbconvert
tornado==6.3.2
    # via
    #   ipykernel
    #   jupyter-client
    #   jupyter-server
    #   nbclassic
    #   nbdime
    #   notebook
    #   terminado
traitlets==5.9.0
    # via
    #   comm
    #   ipykernel
    #   ipython
    #   ipywidgets
    #   jupyter-client
    #   jupyter-core
    #   jupyter-events
    #   jupyter-server
    #   matplotlib-inline
    #   nbclassic
    #   nbclient
    #   nbconvert
    #   nbformat
    #   notebook
uc-micro-py==1.0.2
    # via linkify-it-py
urllib3==2.0.2
    # via requests
wcwidth==0.2.6
    # via prompt-toolkit
webencodings==0.5.1
    # via
    #   bleach
    #   tinycss2
websocket-client==1.5.2
    # via jupyter-server
wheel==0.40.0
    # via sphinx-togglebutton
widgetsnbextension==3.6.4
    # via ipywidgets
zipp==3.15.0
    # via importlib-metadata

# The following packages are considered to be unsafe in a requirements file:
# setuptools
```

</details>

So you can be more condfident that the build in CI/for deployment is using the same packages as your local machine.

You can run `pip-compile` again later to bump versions in `requirements.txt`.

**Comment by bsipocz** [comment]
> So you can be more condfident that the build in CI/for deployment is using the same packages as your local machine.

I suggested tox to do this a few times (e.g https://github.com/alan-turing-institute/the-turing-way/issues/2481#issuecomment-1503888746), I guess I just need to commit to it to do it. We use that approach for various libraries, but also for executable book based repos. E.g. https://github.com/numpy/numpy-tutorials

**Comment by JimMadge** [comment]
I'm not familiar with tox, so forgive me if I'm wrong.

It looks like it solves a different problem, declarative build/test environments and commands, Whereas pip-tools is about maintaining a record of exact dependencies while keeping updating easy.

Both would be valuable. pip-tools doesn't help you specify a Python version or declare multiple environments.
It looks like the tox configuration can point to dependencies from a `requirements.txt` file.

Could we then have a suite of 'easy' commands like

```
tox run py-3-11-build
tox run py-3-10-build
tox run py-3-11-link-check
```

**Comment by HaoZeke** [comment]
I'm not sure how this is better / different from `pdm`?https://pdm.fming.dev/

**Comment by JimMadge** [comment]
`pip-tools`? `pdm` looks like a comprehensive package/project manager like `poetry` or `hatch`. Which might be valuable but I don't think it hurts the project to not use one.

My thought was more to ensure each commit (and hence tag, release) has a full list of pinned dependencies, so that the book at that point can be rebuilt later on, _while_ maintaining a small list of (unpinned) 'direct' dependencies. `pip-tools` does that with just a couple of commands (and one file rename :smile:).

**Comment by JimMadge** [comment]
Closed in #3496
